### ISSUE TITLE
ENH: "vectorize" newton and other scalar methods

### ISSUE THREAD

**Issue by mikofski** [scipy/scipy#8354]
# Problem
Commonly we have to analyze large datasets, like timeseries. If user wants to find zeros of multiple scalar problems simultaneously, they have to use `np.vectorize`, a multidimensional solver, use multiprocessing, threading, a jitter, or custom code. There are some downsides to each of these.
* `np.vectorize` scales linearly with size of inputs, so it gets very slow for large datasets
* IMO multidimensional solvers are overkill, but I should investigate this option more. perhaps using a sparse jacobian will improve performance, but I tried `hybrj` (`root`, `fsolve`) and didn't get fast results with even small datasets compared to `brentq` or `newton`
* multiprocessing, threading, numba, etc. seem like the wrong approach here
* if I customize code, then who will maintain it?

## Proposal
The newton and bisection methods can be adapted easily to handle multiple scalar problems simultaneously. I see tremendous speed up with a simple custom newton code, with no back tracking or line-searching. This might be improved more by implementing in C, using numba, or cython maybe?

A quick comparison of the proposed custom "vectorized" newton versus using `np.vectorize` to wrap the existing one, shows a **20x speedup** for solving a [one-diode solar-cell analog using Shockley's diode equation](https://en.wikipedia.org/wiki/Theory_of_solar_cells#Characteristic_equation) with 100 simultaneous orthogonal points. Of course this proposal can be improved. But I think it would be better to have it in SciPy than many individual users' custom code.

Furthmore, if we [allow the jac to be returned with func, like some other functions already do](./issues/8345), then this halves the time, and we get 40x speedup!

### Reproducing code example:
```
def newton(func, x0, fprime, args=(), tol=1.48e-08, maxiter=50, full_output=None):
    resnorm = np.inf  # nor of residuals
    for iter in range(maxiter):
        fval = func(x0, *args)
        if fprime is True:
            fval, jval = fval
        else:
            jval = fprime(x0, *args)
        resnorm = np.sqrt(np.sum(fval**2))
        if resnorm < tol:
            break
        newton_step = fval / jval
        # do some clever damping here,
        # or use Numerical Recipes in C ch. 9.6-9.7 and do line search
        x0 -= newton_step
    if full_output:
         return x, fval, resnorm, newton_step, iter
    return x0
```

### Error message:
```
~\AppData\Local\Continuum\anaconda3\lib\site-packages\scipy\optimize\zeros.py in newton(func, x0, fprime, args, tol, maxiter, fprime2)
    138             myargs = (p0,) + args
    139             fder = fprime(*myargs)
--> 140             if fder == 0:
    141                 msg = "derivative was zero."
    142                 warnings.warn(msg, RuntimeWarning)

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

### Scipy/Numpy/Python version information:
```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
0.19.1 1.13.3 sys.version_info(major=3, minor=6, micro=3, releaselevel='final', serial=0)
```

**Comment by mikofski** [comment]
I will submit a PR asap

**Comment by person142** [comment]
This came up before: https://github.com/scipy/scipy/issues/7242.

As you can see I was against the idea. Handling all reasonable halting conditions for a scalar solver is already a PITA, and I think the problem gets much worse for a vectorized scalar solver. IMO it is better to provide a Cython API to a scalar solver and let users handle looping over it.

Just me though, maybe others are in favor. :-)

**Comment by ilayn** [comment]
I don't follow the argument very closely I guess but np.vectorize does not vectorize anything. It's just a convenience function to avoid for loop writing annoyance you want to apply a custom function over an array.

Also bisection methods are line searches too anyways. So I don't really get that point too. Care to elaborate?

Also halting conditions must be tested over ill-conditioned problems and I agree they are a major source of pain.

**Comment by andyfaff** [comment]
IIUC this kind of problem is normally tackled by parallel processing, using something like `multiprocessing`, `mpi4py`, or `joblib`.

**Comment by mikofski** [comment]
@andyfaff I had thought of mp, and even threads can be used in this case [since numpy releases the GIL](http://scipy-cookbook.readthedocs.io/items/ParallelProgramming.html#Threads). But I think that may be overkill here, especially since all we need is to use numpy arrays instead of scalars. All of the calculations in both newton and the other scalar zero methods are [ufuncs that can use ndarrays](https://docs.scipy.org/doc/numpy/reference/ufuncs.html) instead of scalars. The only change required would be the convergence criteria - instead of looking at a single value, you'd either need to look at each individually or all together, like rms in my example code.

**Comment by mikofski** [comment]
addressed in PR #8357

**Comment by mikofski** [comment]
@person142 can you explain a bit more about the Cython API option? Not being much of an expert on Cython, I googled it, and I came up with these possibly related links:
* [Cython API for `special` functions](https://docs.scipy.org/doc/scipy/reference/special.cython_special.html) - Are you suggesting that there be something similar to `scipy.special.cython_special` for `optimize`? Seems like a very good idea, but maybe not the solution to this particular problem.
    - IMO this seems like an undo burden on the typical user, who IMO is more likely to just roll their own 11 lines of Python than learn how to use Cython, which I believe most users don't use frequently.
    - Also IMO maintaining a package that uses Cython is more difficult than a pure Python package
    - Also ironically, most of the `scipy.special` methods exposed in the Cython API are _already_ ufuncs, so they don't need to be vectorized.
* [Cython for NumPy Users](https://github.com/cython/cython/wiki/tutorials-numpy), also [here](http://cython.readthedocs.io/en/latest/src/userguide/numpy_tutorial.html) - So assuming that we created a Cython API `scipy.optimize.cython_optimize` then we can import it into our Cython code:

        cimport scipy.optimize.cython_optimize

    this is a super cool idea, except that `newton` is still not a ufunc, so we would still have to loop over it, and I just don't have enough experience with Cython to be sure that its smart enough to compile and take advantage of all the tricks that numpy uses.

Overall, if I'm interpretting you correctly?, I'm excited about your idea of a Cython API for `newton` (and others?) especially if it can utilize compile-time optimizations like SIMD or whatever other tricks there may be. Perhaps SciPy already does this for the pure C implementations in [`Zeros`](../tree/master/scipy/optimize/Zeros), but if not, and Cython does this better (or maybe [Pythran](http://pythran.readthedocs.io/en/latest/)?) then we could extend the Cython API to more `optimize` methods that are not already optimized.

Another route, would be runtime compile optimizations using [Numba/llvm](https://numba.pydata.org/). Do you (or anyone?) think the SciPy community is open to including Numba and [llvmlite](https://llvmlite.readthedocs.io/en/latest/) as dependencies to Scipy?

However, I still believe that making newton a ufunc first will be the most valuable to the majority of typical users. I also believe, at least for newton, that this might be able to be accomplished with minimal effort and minor changes, since it is a simple algorithms, that can be constructed from existing ufuncs.

**Comment by person142** [comment]
> Overall, if I'm interpretting you correctly?

Pretty much yes.

> this is a super cool idea, except that newton is still not a ufunc, so we would still have to loop over it

Right. Admittedly I'm a little tired of vectorizing. Some is great, e.g. `np.sin(x) + np.cos(x)` for `x` an array. But at a certain point I think it leads to more confusing code where having a user write a for loop would be clearer/have a less bloated API. But looping in Python is slow, so my preferred solution is exposing a Cython API so that users can write fast loops. Maybe I've forgotten how high the "learn Cython" overhead is though...

> Do you (or anyone?) think the SciPy community is open to including Numba and llvmlite as dependencies to Scipy?

At this point I think pretty much no because SciPy's dependencies are already pretty heavy.

> Also ironically, most of the scipy.special methods exposed in the Cython API are already ufuncs, so they don't need to be vectorized

The Cython API is designed to solve the problem of using a special function in a tight loop which isn't easily vectorized (maybe it has a lot of branching). Maybe nobody uses it except me though. ;-)

**Comment by mikofski** [comment]
First thanks for your thoughtful responses. I've tried to benchmark a cythonized example of newton to solve a "solar cell" problem for 100,000 different cases and compared them with numpy and mp at the suggestion of @andyfaff (see [this comment](../pull/8357#issuecomment-362849225)). I'm reposting some results here and appending the new cython results.

**TL;DR**: Using NumPy arrays is about 200x faster than using multiprocessing or regular Python loops (_a la_ [`np.vectorize`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.vectorize.html)), and bout 25x faster than my first attempt at cythonizing the newton solar-cell example.

|method|`%timeit`|
|------|-------|
|loop|1 loop, best of 3: 3.73 s per loop|
|mp|1 loop, best of 3: 2.66 s per loop|
|numpy|100 loops, best of 3: 13.9 ms per loop|
|cython|1 loop, best of 3: 350 ms per loop|

Here are the code snippets:
* [multiprocessing](https://pastebin.com/5E3qgxes)
* [numpy arrays in this PR](https://pastebin.com/gaWL580d)
* [cython](https://gist.github.com/mikofski/164f4cf202737d68c6277d4a97c447f2)

IMHO the numpy "vectorized" code in #8357  actually _is_ easier to read, and it required so few changes for it to be vectorized - it actually removes 4 lines of code.

I'm sure my cythonized example code (see [gist](https://gist.github.com/mikofski/164f4cf202737d68c6277d4a97c447f2)) could be vastly improved, but I though there were substantial roadblocks to widespread use:

1. as you said, the `special` cython API were for
    >a tight loop which isn't easily vectorized
    
    ... but in this case, the loops are simple maps

2. I had to [define the callback function statically, which fixed its inputs](https://gist.github.com/mikofski/164f4cf202737d68c6277d4a97c447f2#file-zeros-pxd-L2) as a double `x`, and a tuple `args` which I think would work for most callbacks

3. Therefore the callbacks themselves had to be `cdef` instead of `def` or `cpdef`.

4. I didn't even try to figure out how to handle the different input arguments between the 3 methods: newton, halley, and secant, so I just split them up.

5. little cython quirks like no spat operators _eg_: `*myargs`

I think for truly exceptional performance, we could use cython _and_ numpy.

**Comment by wholmgren** [comment]
@mikofski thanks for the examples. I hacked away at the cython [here](https://github.com/wholmgren/ivcython). It's about 3x faster than when I started (270 ms down to 90 ms for 100k points). My guess is that the remaining bottleneck is the way that the args get passed around as a tuple. Maybe they could be packaged up as structs instead?

In any case, it's *a lot* more work than having a canned vectorized optimizer.

It seems to me that adding support for a vectorized solver doesn't hurt anyone. I think it would be beneficial to a lot of people working on a lot of different problems. If I understand correctly, the vectorized solver will work best when the convergence rate is similar for all points. I'm guessing that describes the majority of use cases.

On the other hand, the non-newton zeros functions are written in c. It seems like turning them into python functions so that they can be vectorized is going the wrong way.

**Comment by thunderfish24** [comment]
I think it would be very valuable (esp. performance-wise, such as for Monte Carlo or long time-series analyses) and not at all "messy" to "vectorize" scipy.optimize.newton() in a pure python implementation. First, this function does not return anything but the supposed zero, and, in particular, it does not return a more complicated RootResults object. Second, the stopping criterion is easily vectorized and is not unreasonably affected by the problem being "harder" for some inputs than others. Granted, this function is somewhat risky to use, but no more risky if it is vectorized. It also seems that this approach would be way easier to implement and maintain, for both producer and consumer.

**Comment by mikofski** [comment]
I also started a [thread on the SciPy-Dev mailing list](https://mail.python.org/pipermail/scipy-dev/2018-February/022423.html) for discussion .
